beetsplug/replaygain.py

```diff
@@ -905,21 +928,55 @@ def handle_track(self, item, write):
 
         self._log.info(u'analyzing {0}', item)
 
-        try:
-            track_gains = self.backend_instance.compute_track_gain([item])
-            if len(track_gains) != 1:
-                raise ReplayGainError(
-                    u"ReplayGain backend failed for track {0}".format(item)
+        self._log.debug(u'format: {0}', item.format)
+
+        if item.format == 'Opus':
+            backend_name = 'bs1770gain'
+            backend_instance = self.backend_instance
+
+            try:
+                self.backend_instance = self.backends[backend_name](
+                    self.config, self._log
                 )
+            except (ReplayGainError, FatalReplayGainError) as e:
+                raise ui.UserError(
+                    u'replaygain initialization failed: {0}'.format(e))
 
-            self.store_track_gain(item, track_gains[0])
-            if write:
-                item.try_write()
-        except ReplayGainError as e:
-            self._log.info(u"ReplayGain error: {0}", e)
-        except FatalReplayGainError as e:
-            raise ui.UserError(
-                u"Fatal replay gain error: {0}".format(e))
+            try:
+                track_gains = \
+                    self.backend_instance.compute_track_r128_gain([item])
+                if len(track_gains) != 1:
+                    raise ReplayGainError(
+                        u"ReplayGain backend failed for track {0}".format(item)
+                    )
+                self.store_track_r128_gain(item, track_gains[0])
+                if write:
+                    item.try_write()
+            except ReplayGainError as e:
+                self._log.info(u"ReplayGain error: {0}", e)
+            except FatalReplayGainError as e:
+                raise ui.UserError(
+                    u"Fatal replay gain error: {0}".format(e))
+
+            self.backend_instance = backend_instance
+
+        else:
+
+            try:
+                track_gains = self.backend_instance.compute_track_gain([item])
+                if len(track_gains) != 1:
+                    raise ReplayGainError(
+                        u"ReplayGain backend failed for track {0}".format(item)
+                    )
+
+                self.store_track_gain(item, track_gains[0])
+                if write:
+                    item.try_write()
+            except ReplayGainError as e:
+                self._log.info(u"ReplayGain error: {0}", e)
+            except FatalReplayGainError as e:
+                raise ui.UserError(
+                    u"Fatal replay gain error: {0}".format(e))
```

Comment:
Hmm; I see what you mean about duplicate structure between the two branches. It does seem like a good idea to trim this down, but I don't have any non-obvious suggestions at the moment…